File: packages/wnode-sdk-ts/test/mesh/discovery.test.ts
import { MeshDiscovery } from '../../src/mesh/discovery';
import { WnodeDeterminismError } from '../../src/errors';
import { PeerInfo } from '../../src/mesh/types';

describe('MeshDiscovery', () => {
  let discovery: MeshDiscovery;

  beforeEach(() => {
    discovery = new MeshDiscovery('node-1', '1.0.0', '1.0');
  });

  it('accepts valid peers', () => {
    const peer: PeerInfo = {
      nodeId: 'node-2',
      sdkVersion: '1.0.0',
      protocolVersion: '1.0',
      strictDeterminism: true,
      capabilities: []
    };

    expect(() => discovery.handleHeartbeat(peer)).not.toThrow();
    expect(discovery.getPeers().length).toBe(1);
  });

  it('rejects mismatched sdkVersion', () => {
    const peer: PeerInfo = {
      nodeId: 'node-2',
      sdkVersion: '0.9.0',
      protocolVersion: '1.0',
      strictDeterminism: true,
      capabilities: []
    };

    expect(() => discovery.handleHeartbeat(peer)).toThrow(WnodeDeterminismError);
  });

  it('rejects unsafe determinism flags', () => {
    const peer: PeerInfo = {
      nodeId: 'node-2',
      sdkVersion: '1.0.0',
      protocolVersion: '1.0',
      strictDeterminism: false,
      capabilities: []
    };

    expect(() => discovery.handleHeartbeat(peer)).toThrow(WnodeDeterminismError);
  });
});

File: archive/root_legacy/wnode-sdk-go/mesh/discovery_test.go
package mesh

import (
	"testing"
	"github.com/wnodeltd/wnode/wnode-sdk-go"
)

func TestMeshDiscovery_AcceptsValidPeers(t *testing.T) {
	discovery := NewMeshDiscovery("node-1", "1.0.0", "1.0")

	peer := PeerInfo{
		NodeID:            "node-2",
		SDKVersion:        "1.0.0",
		ProtocolVersion:   "1.0",
		StrictDeterminism: true,
	}

	err := discovery.HandleHeartbeat(peer)
	if err != nil {
		t.Fatalf("Expected no error, got: %v", err)
	}

	if len(discovery.GetPeers()) != 1 {
		t.Fatalf("Expected 1 peer, got %d", len(discovery.GetPeers()))
	}
}

func TestMeshDiscovery_RejectsMismatchedSDKVersion(t *testing.T) {
	discovery := NewMeshDiscovery("node-1", "1.0.0", "1.0")

	peer := PeerInfo{
		NodeID:            "node-2",
		SDKVersion:        "0.9.0",
		ProtocolVersion:   "1.0",
		StrictDeterminism: true,
	}

	err := discovery.HandleHeartbeat(peer)
	if err == nil {
		t.Fatal("Expected error, got nil")
	}

	detErr, ok := err.(*sdk.WnodeDeterminismError)
	if !ok || detErr.Code != "PEER_REJECTED" {
		t.Fatalf("Expected PEER_REJECTED WnodeDeterminismError, got %v", err)
	}
}

func TestMeshDiscovery_RejectsUnsafeDeterminism(t *testing.T) {
	discovery := NewMeshDiscovery("node-1", "1.0.0", "1.0")

	peer := PeerInfo{
		NodeID:            "node-2",
		SDKVersion:        "1.0.0",
		ProtocolVersion:   "1.0",
		StrictDeterminism: false,
	}

	err := discovery.HandleHeartbeat(peer)
	if err == nil {
		t.Fatal("Expected error, got nil")
	}
}
